File: backend/app/agent/events.py
"""SSE 事件类型定义 — Agent 循环每一步都向客户端推送事件。"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any


@dataclass
class SSEEvent:
    """SSE 事件，通过 SSE 流推送到客户端。"""

    # 事件类型: thinking | tool_call | tool_result | heal_attempt
    #          | budget_update | rca_partial | rca_complete | error
    type: str
    data: dict[str, Any]


def thinking(text: str) -> SSEEvent:
    """Agent 正在推理。"""
    return SSEEvent(type="thinking", data={"text": text})


def tool_call(tool_name: str, params: dict[str, Any]) -> SSEEvent:
    """Agent 调用了一个工具。"""
    return SSEEvent(type="tool_call", data={"tool": tool_name, "params": params})


def tool_result(
    tool_name: str,
    success: bool,
    data: dict[str, Any] | None = None,
    error: str | None = None,
    latency_ms: int = 0,
    cached: bool = False,
) -> SSEEvent:
    """工具执行结果。"""
    return SSEEvent(
        type="tool_result",
        data={
            "tool": tool_name,
            "success": success,
            "data": data,
            "error": error,
            "latency_ms": latency_ms,
            "cached": cached,
        },
    )


def budget_update(
    tokens_used: int, tokens_max: int, calls_used: int, calls_max: int
) -> SSEEvent:
    """预算消耗更新。"""
    return SSEEvent(
        type="budget_update",
        data={
            "tokens_used": tokens_used,
            "tokens_max": tokens_max,
            "calls_used": calls_used,
            "calls_max": calls_max,
        },
    )


def rca_complete(report: str, confidence: str = "high") -> SSEEvent:
    """RCA 报告生成完成。"""
    return SSEEvent(
        type="rca_complete",
        data={"report": report, "confidence": confidence},
    )


def rca_partial(report: str, missing: list[str]) -> SSEEvent:
    """部分 RCA（预算耗尽或证据不足）。"""
    return SSEEvent(
        type="rca_partial",
        data={"report": report, "missing": missing, "confidence": "low"},
    )


def error(message: str) -> SSEEvent:
    """错误事件。"""
    return SSEEvent(type="error", data={"message": message})

File: frontend/src/types/events.ts
// SSE 事件类型定义 — 与后端 backend/app/agent/events.py 对齐

export type SSEEventType =
  | "thinking"
  | "tool_call"
  | "tool_result"
  | "heal_attempt"
  | "playbook_hint"
  | "budget_update"
  | "rca_partial"
  | "rca_complete"
  | "error";

/** self-heal 事件：工具执行失败后自动修正重试 */
export interface HealAttemptEvent extends BaseSSEEvent {
  type: "heal_attempt";
  data: {
    tool: string;
    /** 原始失败查询 */
    original_query: string;
    /** 失败原因 */
    error: string;
    /** 修正后的查询 */
    healed_query: string;
    /** 重试是否成功 */
    success: boolean;
  };
}

export interface BaseSSEEvent {
  type: SSEEventType;
  data: Record<string, unknown>;
}

export interface ThinkingEvent extends BaseSSEEvent {
  type: "thinking";
  data: { text: string };
}

export interface ToolCallEvent extends BaseSSEEvent {
  type: "tool_call";
  data: { tool: string; params: Record<string, unknown> };
}

export interface ToolResultEvent extends BaseSSEEvent {
  type: "tool_result";
  data: {
    tool: string;
    success: boolean;
    data: Record<string, unknown> | null;
    error: string | null;
    latency_ms: number;
    cached: boolean;
  };
}

export interface BudgetUpdateEvent extends BaseSSEEvent {
  type: "budget_update";
  data: {
    tokens_used: number;
    tokens_max: number;
    calls_used: number;
    calls_max: number;
  };
}

export interface RCACompleteEvent extends BaseSSEEvent {
  type: "rca_complete";
  data: { report: string; confidence: "high" | "medium" | "low" };
}

export interface RCAPartialEvent extends BaseSSEEvent {
  type: "rca_partial";
  data: {
    report: string;
    missing: string[];
    confidence: "low";
    suggestions?: string[];
  };
}

export interface ErrorEvent extends BaseSSEEvent {
  type: "error";
  data: { message: string };
}

export interface PlaybookHintEvent extends BaseSSEEvent {
  type: "playbook_hint";
  data: {
    playbook_id: string;
    playbook_name: string;
    score: number;
    matched_keywords: string[];
    steps: { probe: string; query_template: string; purpose: string }[];
    common_root_causes: string[];
  };
}

export type SSEEvent =
  | ThinkingEvent
  | ToolCallEvent
  | ToolResultEvent
  | HealAttemptEvent
  | BudgetUpdateEvent
  | RCACompleteEvent
  | RCAPartialEvent
  | PlaybookHintEvent
  | ErrorEvent;

// 会话状态
export type SessionStatus = "created" | "running" | "completed" | "error";

export interface Session {
  session_id: string;
  message: string;
  status: SessionStatus;
  created_at: number;
  events: SSEEvent[];
  rca_report: string | null;
  rca_confidence: "high" | "medium" | "low" | null;
  is_partial: boolean;
  missing_queries: string[];
  suggestions: string[];
  followup_count: number;
}

// 预算信息
export interface BudgetInfo {
  tokens_used: number;
  tokens_max: number;
  calls_used: number;
  calls_max: number;
}
